src/PublicAllocator.sol

```diff
@@ -1,25 +1,127 @@
+pragma solidity 0.8.24;
 
-import {IMetaMorpho, MarketAllocation} from "../lib/metamorpho/src/interfaces/IMetaMorpho.sol";
+import {
+    Id, IMorpho, IMetaMorpho, MarketAllocation, MarketParams
+} from "../lib/metamorpho/src/interfaces/IMetaMorpho.sol";
 
-contract PublicAllocator {
+import {MarketParamsLib} from "../lib/metamorpho/lib/morpho-blue/src/libraries/MarketParamsLib.sol";
+
+import {MorphoLib} from "../lib/metamorpho/lib/morpho-blue/src/libraries/periphery/MorphoLib.sol";
+
+import {MorphoBalancesLib} from "../lib/metamorpho/lib/morpho-blue/src/libraries/periphery/MorphoBalancesLib.sol";
+
+import {SharesMathLib} from "../lib/metamorpho/lib/morpho-blue/src/libraries/SharesMathLib.sol";
+
+import {Market} from "../lib/metamorpho/lib/morpho-blue/src/interfaces/IMorpho.sol";
+import {UtilsLib} from "./libraries/UtilsLib.sol";
+
+import {ErrorsLib} from "./libraries/ErrorsLib.sol";
+import {EventsLib} from "./libraries/EventsLib.sol";
+import {FlowCap, FlowConfig, SupplyConfig, IPublicAllocatorStaticTyping} from "./interfaces/IPublicAllocator.sol";
+
+contract PublicAllocator is IPublicAllocatorStaticTyping {
+    using MorphoLib for IMorpho;
+    using MorphoBalancesLib for IMorpho;
+    using MarketParamsLib for MarketParams;
+    using SharesMathLib for uint256;
+    using UtilsLib for uint256;
+    using UtilsLib for uint128;
+
+    /// CONSTANTS ///
+
+    address public immutable OWNER;
+    IMorpho public immutable MORPHO;
+    IMetaMorpho public immutable VAULT;
 
     /// STORAGE ///
 
-    address public owner;
-    IMetaMorpho public vault;
+    uint256 public fee;
+    mapping(Id => FlowCap) public flowCap;
+    mapping(Id => uint256) public supplyCap;
+
+    /// MODIFIER ///
+
+    modifier onlyOwner() {
+        if (msg.sender != OWNER) revert ErrorsLib.NotOwner();
+        _;
+    }
 
     /// CONSTRUCTOR ///
 
-    constructor(address newOwner, IMetaMorpho newVault) {
-        owner = newOwner;
-        vault = newVault;
+    constructor(address newOwner, address vault) {
+        if (newOwner == address(0)) revert ErrorsLib.ZeroAddress();
+        if (vault == address(0)) revert ErrorsLib.ZeroAddress();
+        OWNER = newOwner;
+        VAULT = IMetaMorpho(vault);
+        MORPHO = VAULT.MORPHO();
     }
 
     /// PUBLIC ///
 
-    function reallocate(MarketAllocation[] calldata allocations) external {
-        vault.reallocate(allocations);
+    function reallocate(MarketAllocation[] calldata allocations) external payable {
+        if (msg.value != fee) revert ErrorsLib.IncorrectFee(msg.value);
+
+        uint256[] memory assets = new uint256[](allocations.length);
+        for (uint256 i = 0; i < allocations.length; ++i) {
+            // Do not compute interest twice for every market
+            MORPHO.accrueInterest(allocations[i].marketParams);
+            assets[i] = MORPHO.expectedSupplyAssets(allocations[i].marketParams, address(VAULT));
+        }
+
+        VAULT.reallocate(allocations);
```

Comment:
nb: we would be very happy if MM was returning the changes applied to the markets